src/aspire/abinitio/commonline_sdp.py

```diff
@@ -1,30 +1,249 @@
+        # Define and solve the CVXPY problem.
+        # Create a symmetric matrix variable.
+        G = cp.Variable((n, n), symmetric=True)
+        # The operator >> denotes matrix inequality.
+        constraints = [G >> 0]
+        constraints += [cp.trace(A[i] @ G) == b[i] for i in range(3 * self.n_img)]
+        prob = cp.Problem(cp.Minimize(cp.trace(-S @ G)), constraints)
+        prob.solve()
 
-    def estimate(self):
+        return G.value
+
+    def _deterministic_rounding(self, Gram):
         """
-        perform estimation of orientations
+        Deterministic rounding procedure to recover the rotations from the Gram matrix.
+
+        The Gram matrix contains information about the first two columns of every rotation
+        matrix. These columns are extracted and used to form the remaining column of every
+        rotation matrix.
+
+        :param Gram: A 2n_img x 2n_img Gram matrix.
+
+        :return: An n_img x 3 x 3 stack of rotation matrices.
         """
-        pass
+        logger.info("Recovering rotations from Gram matrix.")
+
+        # Obtain top eigenvectors from Gram matrix.
+        d, v = stable_eigsh(Gram, 5)
+        sort_idx = np.argsort(-d)
+        logger.info(f"Top 5 eigenvalues from (rank-3) Gram matrix: {d[sort_idx]}")
 
-    def output(self):
+        # Only need the top 3 eigen-vectors.
+        v = v[:, sort_idx[:3]]
+
+        # According to the structure of the Gram matrix, the first `n_img` rows, denoted v1,
+        # correspond to the linear combination of the vectors R_{i}^{1}, i=1,...,K, that is of
+        # column 1 of all rotation matrices. Similarly, the second `n_img` rows of v,
+        # denoted v2, are linear combinations of R_{i}^{2}, i=1,...,K, that is, the second
+        # column of all rotation matrices.
+        v1 = v[: self.n_img].T
+        v2 = v[self.n_img : 2 * self.n_img].T
+
+        # Use a least-squares method to get A.T*A and a Cholesky decomposition to find A.
+        A = self._ATA_solver(v1, v2)
+
+        # Recover the rotations. The first two columns of all rotation
+        # matrices are given by unmixing V1 and V2 using A. The third
+        # column is the cross product of the first two.
+        r1 = np.dot(A.T, v1)
+        r2 = np.dot(A.T, v2)
+        r3 = np.cross(r1, r2, axis=0)
+
+        rotations = np.empty((self.n_img, 3, 3), dtype=self.dtype)
+        rotations[:, :, 0] = r1.T
+        rotations[:, :, 1] = r2.T
+        rotations[:, :, 2] = r3.T
+        # Make sure that we got rotations by enforcing R to be
+        # a rotation (in case the error is large)
+        u, _, v = np.linalg.svd(rotations)
+        np.einsum("ijk, ikl -> ijl", u, v, out=rotations)
+
+        return rotations
+
+    @staticmethod
+    def _ATA_solver(v1, v2):
         """
-        Output the 3D orientations
+        Uses a least squares method to solve for the linear transformation A
+        such that A*v1=R1 and A*v2=R2 are the columns of the rotations matrices.
```

Comment:
“correspond to the first and second columns of a sequence of rotation matrices” (since we don't know the actual rotation matrices).

Reply:
Sounds good. Done!